src/components/application_manager/include/application_manager/system_time/system_time_handler_impl.h

```diff
@@ -0,0 +1,151 @@
+ public:
+  /**
+   * @brief SystemTimeHandlerImpl creates the instance of the class.
+   */
+  explicit SystemTimeHandlerImpl(ApplicationManager& application_manager);
+
+  /**
+   * @brief ~SystemTimeHandlerImpl
+   */
+  ~SystemTimeHandlerImpl();
+
+ private:
+  /**
+   * @brief on_event allows to process certain events from the system.
+   * Event which are handles within this methods  are: OnSystemTimeReady
+   * in order to send system time query and GetSystemTimeResponse in order
+   * to retrieve utc time and notify all the listeners with new time value.
+   */
+  void on_event(const application_manager::event_engine::Event& event) FINAL;
+
+  /**
+   * @brief DoSystemTimeQuery sends the appropriate request to the system
+   * and subscribes the class to the response. This is asynchronous request
+   * thus it won't block until the system respond and returns immediately.
+   */
+  void DoSystemTimeQuery() FINAL;
+
+  /**
+   * @brief DoSubscribe allows to subscribe certain listener for
+   * SystemTime. This certain implementation does not maintain
+   * listeners collection instead it save the pointer to listener.
+   * @note the class is not reponsible for the pointer's lifecycle.
+   * So consider to explicitly unsibsscribe from event when listener
+   * is going to be destroyed.
+   * @param listener the listener which will be notified
+   * in case of SystemTime appeared.
+   */
+  void DoSubscribe(utils::SystemTimeListener* listener) FINAL;
+
+  /**
+   * @brief DoUnsubscribe assigns the pointer to listener to NULL.
+   * This certain implementation ignores passed parameter.
+   * @param listener pointer to the object which should be
+   * unsubscribed from events.
+   */
+  void DoUnsubscribe(utils::SystemTimeListener* listener) FINAL;
+
+  /**
+   * @brief FetchSystemTime this method allows to obtain
+   * recently received utc time.
+   * @note it is up to user to check whether the returned
+   * argument is valid
+   * @return the time value returned by system.
+   */
+  time_t FetchSystemTime() FINAL;
+
+  /**
+   * @brief SendTimeRequest sends the request to the system
+   * and subscribes for response.
+   */
+  void SendTimeRequest();
+
+  /**
+   * @brief OnSystemTimeResponse allows to process GetSystemTimeResponse
+   * @param event contains response parameters aka "systemTime".
+   */
+  void OnSystemTimeResponse(
+      const application_manager::event_engine::Event& event);
+
+  /**
+   * @brief OnSystemTimeReady allows to process OnSystemTimeready notification
+   * from HMI. It unsubscribes from the mentioned notification and sends
+   * GetSystemTimeRequest to HMI in order to obtain system time
+   */
+  void OnSystemTimeReady();
+
+  /**
+   * @brief Checks if UTC time is ready
+   * and can be requested by GetSystemTime request
+   * @return True if HMI is ready to provide UTC time
+   * otherwise False
+   */
+  bool is_utc_time_ready() const FINAL;
```

Comment:
@AByzhynar not clear what does tit means ` HMI is ready to provide UTC time`. Please describe, what exactly can pretend HMI to provide get system Time. Also From method name developer could understand that HMI already provided system time to SDL and he can call FetchSystemTime method to get it.